---
title: F. ChIP-seq
author: Martin Morgan (mtmorgan@fredhutch.org)
date: "`r Sys.Date()`"
output:
  BiocStyle::html_document:
    markdown_strict: true
    toc: true
vignette: >
  %\VignetteIndexEntry{F. ChIP-seq}
  %\VignetteEngine{knitr::rmarkdown}
  \usepackage[utf8]{inputenc}
---

```{r style, echo=FALSE, results='asis'}
BiocStyle::markdown()
```


```{r setup, echo=FALSE, cache=TRUE}
options(digits=3)
suppressPackageStartupMessages({
    requireNamespace("genefilter", quietly=TRUE)
})
```

# Motivation: differential binding

Key references

- Kharchenko, Tolstorukov, and Park
  ([2008](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2597701/)).
- Lun and Smyth ([2014](http://dx.doi.org/10.1093/nar/gku351)).

# Work flow & statistical issues

## ChIP-seq

Kharchenko et
al. ([2008](http://www.ncbi.nlm.nih.gov/pmc/articles/PMC2597701/)).
![ChIP-seq Overview](our_figures/ChIPSeq_nbt-1508-F1.jpg)

- Tags versus sequenced reads; single-end read extension in 3'
  direction
- Strand shift / cross-correlation
- Defined (narrow, e.g., transcription factor binding sites) versus
  diffuse (e.g., histone marks) peaks

ChIP-seq for differential binding

- Designed experiment with replicate samples per treatment
- Analysis using insights from microarrays / RNA-seq

Novel statistical issues

- Inferring peaks without 'data snooping' (using the same data twice,
  once to infer peaks, once to estimate differential binding)
- Retaining power
- Minimizing false discovery rate

## Peak calling

'Known' ranges

- Count tags in pre-defined ranges, e.g., promoter regions of known
  genes
- Obvious limitations, e.g., regulatory elements not in specified
  ranges; specified range contains multiple regulatory elements with
  complementary behavior

_de novo_ windows

- Width: narrow peaks, 1bp; broad peaks, 150bp
- Offset: 25-100bp; influencing computational burden

_de novo_ peak calling

- Third-party software (many available;
  [MACS](http://liulab.dfci.harvard.edu/MACS/) commonly used)
- Various strategies for calling peaks -- Lun & Smyth,
  [Table 1](http://nar.oxfordjournals.org/content/42/11/e95/T1.expansion.html)

    - Call each sample independently; intersection or union of peaks
      across samples, ...
    - Call peaks from a pooled library
    - ...

- How to choose? -- Lun & Smyth,
  
    - Under the null hypothesis, type I error rate is uniform
    - [Table 2](http://nar.oxfordjournals.org/content/42/11/e95/T2.expansion.html):
      consequences for type I error
    - Best strategy: call peaks from a pooled library

```{r null-p, cache=TRUE}
## 100,000 t-tests under the null, n = 6
n <- 6; m <- matrix(rnorm(n * 100000), ncol=n)
P <- genefilter::rowttests(m, factor(rep(1:2, each=3)))$p.value
quantile(P, c(.001, .01, .05))
hist(P, breaks=20)
```

_de novo_ hybrid strategies

## Differential binding (`r Biocpkg("csaw")`)

## Annotation (`r Biocpkg("ChIPseeker")`)

# Practical: [csaw](http://bioconductor.org/packages/devel/bioc/html/csaw.html)

# Practical Annotation
